makefile

```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+# variables
+COVERAGE_DIR='coverage_data'
+TEST_DIR=tests/unit
+
+.PHONY: createVirtualenv
+createVirtualenv:
+	python3 -m pip install --user virtualenv
+	python3 -m venv env
+	source env/bin/activate
+
+########
+# TEST #
+########
+.PHONY: preInstall
+preInstall:
+	pip install -U pytest
+	pytest --version
+
+.PHONY: test
+test:
+	pytest $(TEST_DIR)
+	@echo 'Tests Completed'
+
+.PHONY: help
+help:
+	@echo 'List all the make tergets available for use with "make list"'
+
+.PHONY: list
+list:
+	@$(MAKE) -pRrq -f $(lastword $(MAKEFILE_LIST)) : 2>/dev/null | awk -v RS= -F: '/^# File/,/^# Finished Make data base/ {if ($$1 !~ "^[#.]") {print $$1}}' | sort | egrep -v -e '^[^[:alnum:]]' -e '^$@$$'
```

Comment:
I still don't understand what it means. Why does this line need to be so complex?

Reply:
So what this does it just lists all the targets for the makefile, you can find more details here: [Targets for makefile](https://stackoverflow.com/questions/4219255/how-do-you-get-the-list-of-targets-in-a-makefile)